src/legacy/shrink.rs

```diff
@@ -0,0 +1,396 @@
+    pub fn create_new() -> [Self; MAX_CODE + 1] {
+        let mut codetab = (0..=u8::MAX)
+            .map(|i| Codetab {
+                prefix_code: Some(i as u16),
+                ext_byte: i,
+                len: 1,
+                last_dst_pos: 0,
+            })
+            .collect::<Vec<_>>();
+        codetab.resize(MAX_CODE + 1, Codetab::default());
+        codetab.try_into().unwrap()
+    }
+}
+
+fn unshrink_partial_clear(codetab: &mut [Codetab], queue: &mut CodeQueue) {
+    let mut is_prefix = [false; MAX_CODE + 1];
+
+    // Scan for codes that have been used as a prefix.
+    for i in CONTROL_CODE + 1..=MAX_CODE {
+        if let Some(prefix_code) = codetab[i].prefix_code {
+            is_prefix[prefix_code as usize] = true;
+        }
+    }
+
+    // Clear "non-prefix" codes in the table; populate the code queue.
+    let mut code_queue_size = 0;
+    for i in CONTROL_CODE + 1..MAX_CODE {
+        if !is_prefix[i] {
+            codetab[i].prefix_code = None;
+            queue.codes[code_queue_size] = Some(i as u16);
+            code_queue_size += 1;
+        }
+    }
+    queue.codes[code_queue_size] = None; // End-of-queue marker.
+    queue.next_idx = 0;
+}
+
+/// Read the next code from the input stream and return it in next_code. Returns
+/// false if the end of the stream is reached. If the stream contains invalid
+/// data, next_code is set to INVALID_CODE but the return value is still true.
+fn read_code<T: std::io::Read, E: Endianness>(
+    is: &mut BitReader<T, E>,
+    code_size: &mut u8,
+    codetab: &mut [Codetab],
+    queue: &mut CodeQueue,
+) -> io::Result<Option<u16>> {
+    // assert(sizeof(code) * CHAR_BIT >= *code_size);
+    let code = is.read::<u16>(*code_size as u32)?;
+
+    // Handle regular codes (the common case).
+    if code != CONTROL_CODE as u16 {
+        return Ok(Some(code));
+    }
+
+    // Handle control codes.
+    if let Ok(control_code) = is.read::<u16>(*code_size as u32) {
+        match control_code {
+            INC_CODE_SIZE => {
+                if *code_size >= MAX_CODE_SIZE {
+                    return Err(io::Error::new(
+                        io::ErrorKind::InvalidData,
+                        "tried to increase code size when already at maximum",
+                    ));
+                }
+                *code_size += 1;
+            }
+            PARTIAL_CLEAR => {
+                unshrink_partial_clear(codetab, queue);
+            }
+            _ => {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!("Invalid control code {}", control_code),
+                ));
+            }
+        }
+        return read_code(is, code_size, codetab, queue);
+    }
+    Ok(None)
+}
+
+/// Output the string represented by a code into dst at dst_pos. Returns
+/// HWUNSHRINK_OK on success, and also updates *first_byte and *len with the
+/// first byte and length of the output string, respectively.
```

Comment:
Update this comment, and since `first_byte` and `len` are write-only and implement `Copy`, make them part of the `Ok` return tuple instead.

Reply:
fixed. Len isn't needed anymore - VecDeque knows it's length :).